peps/pep-0759.rst

```diff
@@ -0,0 +1,472 @@
+- External wheel hosts **SHOULD** use a content delivery network (`CDN
+  <https://en.wikipedia.org/wiki/Content_delivery_network>`__), just as PyPI does.
+- External wheel hosts **MUST** commit to a stable URL for all wheels they host.
+- Externally hosted wheels **MUST NOT** be removed from an external wheel host unless the
+  corresponding ``.rim`` file is deleted from PyPI first, and **MUST NOT** remove external
+  wheels for yanked releases.
+- External wheel hosts **MUST** support `HTTP range requests`_.
+- External wheel hosts **SHOULD** support the `HTTP/2`_ protocol.
+
+Security
+========
+
+Several factors as described in this proposal should mitigate security
+concerns with externally hosted wheels, such as:
+
+- Wheel file checksums **MUST** be included in ``.rim`` files, and once uploaded cannot be
+  changed. Since the checksum stored on PyPI is immutable and required, it is not possible
+  to spoof an external wheel file, even if the owning organization lost control of their
+  hosting domain.
+- Externally hosted wheels **MUST** be served over HTTPS.
+- In order to serve externally hosted wheels, organizations **MUST** be approved by the
+  PyPI admins.
+
+When users identify malware or vulnerabilities in PyPI-hosted projects, they can now
+report this using the `malware reporting facilities <https://pypi.org/security/>`__ on
+PyPI, as also described in this `blog post`_.  The same process can be used to report
+security issues in externally hosted wheels, and the same remediation process should be
+used.  In addition, since organizations with external hosting enabled MUST provide a
+support contact URI, that URI can be used in some cases to report the security issue to
+the hosting organization.  Such organization reporting won't make sense for malware, but
+could indeed be a very useful way to report security vulnerabilities in externally hosted
+wheels.
+
+Rejected ideas
+==============
+
+Several ideas were considered and rejected.
+
+- Requiring digital signatures on externally hosted wheel files, either in
+  addition to or other than hashes. We deem this unnecessary since the
+  checksum requirement should be enough to validate that the metadata on PyPI
+  for a wheel exactly matches the downloaded wheel.  The added complexity of
+  key management outweighs any additional benefit such digital signatures
+  might convey.
+- Hash verification on ``.rim`` file uploads. PyPI *could* verify that the
+  hash in the uploaded ``.rim`` file matches the externally hosted wheel
+  before it accepts the upload, but this requires downloading the external
+  wheel and performing the checksum. This would impose an ordering restriction
+  on uploads, such that the external wheel file is uploaded and publicly
+  available *before* uploading the ``.rim`` file to PyPI. While this might be
+  a good idea to avoid any race condition, it should not be a
+  requirement. This also means that the upload of the ``.rim`` file cannot be
```

Comment:
It now is a requirement, per the *Specification*:

> Externally hosted wheels **MUST** be available before the corresponding ``.rim`` file is
> uploaded to PyPI, otherwise a publishing race condition is introduced, although this
> requirement **MAY** be relaxed for ``.rim`` files uploaded to a :pep:`694` staged release.

Reply:
Good catch!